### PR TITLE
Allow adversaries to run at any layer.

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

We initially assumed an adversary module (e.g. `input_adv_test`) would be the first layer of a model. But we try to loose this assumption so that we can perform data transformation in a model before attack.

This PR allows an adversary module to be placed in any layer of a model, by including `sequence` of current and remaining modules in `**kwargs`.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/attack/adversary.py [hunk 1]
```diff
@@ -24,31 +24,31 @@ class AdversaryCallbackHookMixin(Callback):
 
     callbacks = {}
 
-    def on_run_start(self, adversary, input, target, model, **kwargs) -> None:
+    def on_run_start(self, adversary, **kwargs) -> None:
```
Comment (dxoigmn) [inline mart/attack/adversary.py:27]: I would add type annotations to this mixin.

Comment (dxoigmn) [reply mart/attack/adversary.py:27]: It's unclear to me why you even need `adversary` as an explicit parameter.

File: mart/attack/adversary.py [hunk 2]
```diff
@@ -142,14 +142,15 @@ def on_run_start(self, adversary, input, target, model, **kwargs):
         # We could be at the inference/no-grad mode here.
         # Initialize lazy module
         # FIXME: Perturbers can just use on_run_start/on_run_end to initialize
-        self.perturber(input, target)
+        target = kwargs["target"]
+        self.perturber(self.benign_input, target)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:146]: Where does `self.benign_input` come from?!

File: mart/attack/adversary.py [hunk 3]
```diff
@@ -132,8 +132,8 @@ def done(self) -> bool:
 
         return False
 
-    def on_run_start(self, adversary, input, target, model, **kwargs):
-        super().on_run_start(adversary, input, target, model, **kwargs)
+    def on_run_start(self, adversary, **kwargs):
```
Comment (dxoigmn) [inline mart/attack/adversary.py:135]: I think you should be explicit with the parameters here. That is, explicit define what parameters you need rather than hiding them in `kwargs`. This will catch more bugs.

File: mart/attack/adversary.py [hunk 4]
```diff
@@ -275,14 +264,24 @@ def forward(
...
         **kwargs
     ):
+        # Backward compatible: earlier configs have two unnamed arguments [input, target].
+        kwargs["input"] = input
+        kwargs["target"] = target
+        kwargs["model"] = model
+
+        # The benign input to an adversary is not necessarily the model input.
+        # It could be activations of other modules.
+        self.benign_input = kwargs[self.benign_input_key]
```
Comment (dxoigmn) [inline mart/attack/adversary.py:274]: Why is this necessary? I don't quite understand the comment. What is the "model input" in the context of an "Adversary"? This definitely needs more explanation or rethinking.

File: mart/attack/adversary.py [hunk 5]
```diff
@@ -275,14 +264,24 @@ def forward(
...
+
         # Generate a perturbation only if we have a model. This will update
         # the parameters of self.perturber.
         if model is not None:
-            super().forward(input, target, model, **kwargs)
+            super().forward(**kwargs)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:279]: I wouldn't abuse `**kwargs` this way. Just be explicit.

File: mart/attack/adversary.py (hunk 4 above)
Comment (dxoigmn) [reply mart/attack/adversary.py:274]: Like what is the distinction between benign input and input?

File: mart/attack/adversary.py [hunk 6]
```diff
@@ -275,14 +264,24 @@ def forward(
...
 
         # Get perturbation and apply threat model
-        perturbation = self.perturber(input, target)
-        output = self.threat_model(input, target, perturbation, **kwargs)
+        # The mask projector in perturber may require information from target.
+        perturbation = self.perturber(self.benign_input, target)
+        output = self.threat_model(self.benign_input, target, perturbation)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:284]: I'm hoping this goes away but we shall see. If you can think of a way to rewrite any of the above so that this change does not need to be made, then that is a better solution because you don't break the abstraction. A one off solution like `self.benign_input` is a smell.

File: mart/attack/callbacks/base.py [hunk 7]
```diff
@@ -8,9 +8,7 @@
 from __future__ import annotations
 
 import abc
-from typing import TYPE_CHECKING, Any, Dict, Union
-
-import torch
+from typing import TYPE_CHECKING
```
Comment (dxoigmn) [inline mart/attack/callbacks/base.py:12]: I would revert all these changes and make the callbacks keyword only parameters.

**Review by dxoigmn — CHANGES_REQUESTED** [review]
See comments. I'm still at a loss for why `benign_input` exists.

File: mart/attack/adversary.py [hunk 8]
```diff
@@ -24,31 +24,32 @@ class AdversaryCallbackHookMixin(Callback):
...
     callbacks = {}
 
-    def on_run_start(self, adversary, input, target, model, **kwargs) -> None:
+    def on_run_start(self, **kwargs) -> None:
         """Prepare the attack loop state."""
         for _name, callback in self.callbacks.items():
             # FIXME: Skip incomplete callback instance.
-            callback.on_run_start(adversary, input, target, model, **kwargs)
+            # Give access of self to callbacks by `adversary=self`.
+            callback.on_run_start(adversary=self, **kwargs)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:32]: This is a big logic change and a bad thing to do. The callee should supply `adversary=self` since this is a mixin.

File: mart/attack/adversary.py [hunk 9]
```diff
@@ -132,8 +133,8 @@ def done(self) -> bool:
 
         return False
 
-    def on_run_start(self, adversary, input, target, model, **kwargs):
-        super().on_run_start(adversary, input, target, model, **kwargs)
+    def on_run_start(self, **kwargs):
+        super().on_run_start(**kwargs)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:137]: I would revert this change. Having explicit arguments is good here.

File: mart/attack/adversary.py [hunk 10]
```diff
@@ -142,14 +143,16 @@ def on_run_start(self, adversary, input, target, model, **kwargs):
...
         self.perturber(input, target)
 
         # Split param groups by input elements, so that we can schedule optimizers individually.
         param_groups = [{"params": [param]} for param in self.perturber.parameters()]
         self.opt = self.optimizer_fn(param_groups)
 
-    def on_run_end(self, adversary, input, target, model, **kwargs):
-        super().on_run_end(adversary, input, target, model, **kwargs)
+    def on_run_end(self, **kwargs):
+        super().on_run_end(**kwargs)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:155]: Revert and use explicit arguments.

File: mart/attack/adversary.py [hunk 11]
```diff
@@ -162,51 +165,44 @@ def forward(
...
         """_summary_
 
         Args:
-            input (_type_): _description_
             target (_type_): _description_
             model (_type_): _description_
         """
 
-        self.on_run_start(self, input, target, model, **kwargs)
+        self.on_run_start(input=input, target=target, model=model, **kwargs)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:178]: This is where you should pass `adversary=self`.

Comment (dxoigmn) [reply mart/attack/adversary.py:178]: Same for below.

File: mart/attack/adversary.py [hunk 12]
```diff
@@ -162,51 +165,44 @@ def forward(
...
     @torch.enable_grad()
     @torch.inference_mode(False)
-    def examine(
-        self,
-        input: Union[torch.Tensor, tuple],
-        target: Union[torch.Tensor, Dict[str, Any], tuple],
-        model: torch.nn.Module,
-        **kwargs
-    ):
+    def examine(self, input=None, target=None, model=None, **kwargs):
```
Comment (dxoigmn) [inline mart/attack/adversary.py:205]: Can you revert this and keep the type annotations but as Optional?

File: mart/attack/adversary.py [hunk 13]
```diff
@@ -238,13 +234,7 @@ def examine(
...
     @torch.enable_grad()
     @torch.inference_mode(False)
-    def advance(
-        self,
-        input: Union[torch.Tensor, tuple],
-        target: Union[torch.Tensor, Dict[str, Any], tuple],
-        model: torch.nn.Module,
-        **kwargs
-    ):
+    def advance(self):
```
Comment (dxoigmn) [inline mart/attack/adversary.py:237]: Why did this change?

File: mart/attack/adversary.py [hunk 14]
```diff
@@ -273,16 +263,17 @@ def forward(
         input: Union[torch.Tensor, tuple],
         target: Union[torch.Tensor, Dict[str, Any], tuple],
         model: Optional[torch.nn.Module] = None,
-        **kwargs
+        **kwargs,
```
Comment (dxoigmn) [inline mart/attack/adversary.py:266]: unnecessary (unless make style did this)

File: mart/attack/callbacks/base.py [hunk 15]
```diff
@@ -23,60 +23,60 @@ class Callback(abc.ABC):
...
     def on_run_start(
         self,
-        adversary: Adversary,
-        input: torch.Tensor | tuple,
-        target: torch.Tensor | dict[str, Any] | tuple,
-        model: torch.nn.Module,
+        adversary: Adversary = None,
+        input: torch.Tensor | tuple = None,
+        target: torch.Tensor | dict[str, Any] | tuple = None,
+        model: torch.nn.Module = None,
```
Comment (dxoigmn) [inline mart/attack/callbacks/base.py:29]: Mark Optional? Same with changes below.

File: mart/nn/nn.py [hunk 16]
```diff
@@ -94,15 +94,29 @@ def parse_sequence(self, sequence):
 
     def forward(self, step=None, **kwargs):
         # Try to fetch the customized architectural graph.
-        sequence = self._sequences[step]
+        if "sequence" not in kwargs:
+            # Backward compatible. We may get rid of step in the future.
+            sequence = self._sequences[step]
+        else:
+            sequence = kwargs["sequence"]
```
Comment (dxoigmn) [inline mart/nn/nn.py]: `sequence = kwargs.pop("sequence, self._sequences[step])`?

File: mart/nn/nn.py [hunk 17]
```diff
@@ -94,15 +94,29 @@ def parse_sequence(self, sequence):
...
+
+        # Make a copy of sequence, because it will be destructed in the while loop.
+        sequence = sequence.copy()
+
+        while len(sequence) > 0:
+            # Don't pop the first element yet, because it may be used to re-evaluate the model.
+            key, module = next(iter(sequence.items()))
+            kwargs["sequence"] = sequence
```
Comment (dxoigmn) [inline mart/nn/nn.py]: I don't like manipulating `kwargs` like this. It makes it hard to keep track of what is in `kwargs` as any given time. Instead, I would pass `sequence=sequence` where ever you need it.

**Review by dxoigmn — CHANGES_REQUESTED** [review]
Some more comments. `mart/nn/nn.py` looks good.

File: mart/attack/adversary.py (hunk 8 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: OK. No reference to self in mixin.

File: mart/attack/adversary.py (hunk 9 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Fixed.

File: mart/attack/adversary.py (hunk 10 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Fixed.

File: mart/attack/adversary.py (hunk 12 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Done.

File: mart/attack/adversary.py (hunk 13 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Reverted.

File: mart/attack/adversary.py (hunk 14 above)
Comment (mzweilin) [reply mart/attack/adversary.py:310]: Yes, it's enforced by pre-commit.

File: mart/attack/callbacks/base.py (hunk 15 above)
Comment (mzweilin) [reply mart/attack/callbacks/base.py]: Fixed.

File: mart/nn/nn.py (hunk 17 above)
Comment (mzweilin) [reply mart/nn/nn.py]: Fixed.

File: mart/nn/nn.py (hunk 16 above)
Comment (mzweilin) [reply mart/nn/nn.py]: I made sequence an explicit keyword argument.

File: mart/attack/callbacks/progress_bar.py [hunk 18]
```diff
@@ -15,10 +15,11 @@
 class ProgressBar(Callback):
     """Display progress bar of attack iterations with the gain value."""
 
-    def on_run_start(self, adversary, input, target, model, **kwargs):
+    def on_run_start(self, adversary=None, **kwargs):
         self.pbar = tqdm.tqdm(total=adversary.max_iters, leave=False, desc="Attack", unit="iter")
```
Comment (dxoigmn) [inline mart/attack/callbacks/progress_bar.py:19]: Instead of doing `adversary=None`, I would do:
```
def on_run_start(self, *, adversary, **kwargs):
```
That lets the reader know adversary is a required keyword argument.

File: mart/attack/callbacks/progress_bar.py [hunk 19]
```diff
@@ -15,10 +15,11 @@
...
 
-    def on_examine_end(self, adversary, input, target, model, **kwargs):
+    def on_examine_end(self, adversary=None, **kwargs):
+        input = kwargs["input"]
```
Comment (dxoigmn) [inline mart/attack/callbacks/progress_bar.py:22]: Same comment as above, but also add `input` to the parameter list as a required keyword argument.

File: mart/attack/callbacks/visualizer.py [hunk 20]
```diff
@@ -27,8 +26,8 @@ def __init__(self, folder):
         if not os.path.isdir(self.folder):
             os.makedirs(self.folder)
 
-    def on_run_end(self, adversary, input, target, model, **kwargs):
-        adv_input = adversary(input, target, model=None, **kwargs)
+    def on_run_end(self, adversary=None, input=None, target=None, model=None, **kwargs):
+        adv_input = adversary(input=input, target=target, model=None, **kwargs)
```
Comment (dxoigmn) [inline mart/attack/callbacks/visualizer.py:30]: Use required keyword arguments.

File: mart/attack/callbacks/base.py [hunk 21]
```diff
@@ -23,60 +23,60 @@ class Callback(abc.ABC):
...
     def on_run_end(
         self,
-        adversary: Adversary,
-        input: torch.Tensor | tuple,
-        target: torch.Tensor | dict[str, Any] | tuple,
-        model: torch.nn.Module,
+        adversary: Adversary | None = None,
+        input: torch.Tensor | tuple | None = None,
+        target: torch.Tensor | dict[str, Any] | tuple | None = None,
+        model: torch.nn.Module | None = None,
```
Comment (dxoigmn) [inline mart/attack/callbacks/base.py:79]: I think you can revert all of these changes since the arguments ought to be required. However, you should make these required keyword arguments.

File: mart/attack/adversary.py [hunk 22]
```diff
@@ -240,10 +273,11 @@ def examine(
...
     def advance(
         self,
-        input: Union[torch.Tensor, tuple],
-        target: Union[torch.Tensor, Dict[str, Any], tuple],
-        model: torch.nn.Module,
-        **kwargs
+        adversary: Optional[torch.nn.Module] = None,
+        input: Optional[Union[torch.Tensor, tuple]] = None,
+        target: Optional[Union[torch.Tensor, Dict[str, Any], tuple]] = None,
+        model: Optional[torch.nn.Module] = None,
```
Comment (dxoigmn) [inline mart/attack/adversary.py:282]: Instead of `= None` make these required keyword arguments.

File: mart/attack/adversary.py [hunk 23]
```diff
@@ -202,10 +234,11 @@ def forward(
...
     def examine(
         self,
-        input: Union[torch.Tensor, tuple],
-        target: Union[torch.Tensor, Dict[str, Any], tuple],
-        model: torch.nn.Module,
-        **kwargs
+        adversary: Optional[torch.nn.Module] = None,
+        input: Optional[Union[torch.Tensor, tuple]] = None,
+        target: Optional[Union[torch.Tensor, Dict[str, Any], tuple]] = None,
+        model: Optional[torch.nn.Module] = None,
```
Comment (dxoigmn) [inline mart/attack/adversary.py:244]: Instead of `= None` make these required keyword arguments.

File: mart/attack/adversary.py [hunk 24]
```diff
@@ -162,38 +179,53 @@ def forward(
         self,
         input: Union[torch.Tensor, tuple],
         target: Union[torch.Tensor, Dict[str, Any], tuple],
-        model: torch.nn.Module,
-        **kwargs
+        model: Optional[torch.nn.Module] = None,
```
Comment (dxoigmn) [inline mart/attack/adversary.py:182]: Is `model` really optional? Or should this be be a required keyword argument?

File: mart/attack/adversary.py [hunk 25]
```diff
@@ -148,8 +158,15 @@ def on_run_start(self, adversary, input, target, model, **kwargs):
...
         self.opt = self.optimizer_fn(param_groups)
 
-    def on_run_end(self, adversary, input, target, model, **kwargs):
-        super().on_run_end(adversary, input, target, model, **kwargs)
+    def on_run_end(
+        self,
+        adversary: Optional[torch.nn.Module] = None,
+        input: Optional[Union[torch.Tensor, tuple]] = None,
+        target: Optional[Union[torch.Tensor, Dict[str, Any], tuple]] = None,
+        model: Optional[torch.nn.Module] = None,
```
Comment (dxoigmn) [inline mart/attack/adversary.py:166]: Make these required keyword arguments.

File: mart/attack/adversary.py [hunk 26]
```diff
@@ -132,8 +133,17 @@ def done(self) -> bool:
...
         return False
 
-    def on_run_start(self, adversary, input, target, model, **kwargs):
-        super().on_run_start(adversary, input, target, model, **kwargs)
+    def on_run_start(
+        self,
+        adversary: Optional[torch.nn.Module] = None,
+        input: Optional[Union[torch.Tensor, tuple]] = None,
+        target: Optional[Union[torch.Tensor, Dict[str, Any], tuple]] = None,
+        model: Optional[torch.nn.Module] = None,
```
Comment (dxoigmn) [inline mart/attack/adversary.py:141]: Make these required keyword arguments.

File: tests/test_visualizer.py [hunk 27]
```diff
@@ -28,7 +28,8 @@ def perturb(input, target, model):
     adversary = Mock(spec=Adversary, side_effect=perturb)
 
     visualizer = PerturbedImageVisualizer(folder)
-    visualizer.on_run_end(adversary, input_list, target_list, model)
+    kwargs = {"input": input_list, "target": target_list, "model": model}
+    visualizer.on_run_end(adversary, **kwargs)
```
Comment (dxoigmn) [inline tests/test_visualizer.py:32]: Get rid of `kwargs` and just call the function normally with keyword arguments.

**Review by dxoigmn — CHANGES_REQUESTED** [review]
Instead of abusing `argument=None` make the required keyword arguments where necessary. Any parameters after the '*' parameter become required keyword/named arguments.

File: mart/attack/callbacks/progress_bar.py (hunk 18 above)
Comment (mzweilin) [reply mart/attack/callbacks/progress_bar.py:19]: Thanks. I wasn't aware of such a feature in Python.

File: tests/test_visualizer.py (hunk 27 above)
Comment (mzweilin) [reply tests/test_visualizer.py:32]: Fixed.

File: mart/attack/adversary.py (hunk 26 above)
Comment (mzweilin) [reply mart/attack/adversary.py:141]: Fixed.

File: mart/attack/adversary.py (hunk 25 above)
Comment (mzweilin) [reply mart/attack/adversary.py:166]: Fixed.

File: mart/attack/adversary.py (hunk 24 above)
Comment (mzweilin) [reply mart/attack/adversary.py:182]: `model` is a required keyword argument.

File: mart/attack/adversary.py (hunk 23 above)
Comment (mzweilin) [reply mart/attack/adversary.py:244]: Fixed.

File: mart/attack/adversary.py (hunk 22 above)
Comment (mzweilin) [reply mart/attack/adversary.py:282]: Fixed.

File: mart/attack/callbacks/base.py (hunk 21 above)
Comment (mzweilin) [reply mart/attack/callbacks/base.py:79]: Fixed.

File: mart/attack/callbacks/progress_bar.py (hunk 19 above)
Comment (mzweilin) [reply mart/attack/callbacks/progress_bar.py:22]: Fixed.

File: mart/attack/callbacks/progress_bar.py (hunk 18 above)
Comment (mzweilin) [reply mart/attack/callbacks/progress_bar.py:19]: Fixed.

File: mart/attack/adversary.py [hunk 28]
```diff
@@ -132,8 +133,18 @@ def done(self) -> bool:
...
 
-    def on_run_start(self, adversary, input, target, model, **kwargs):
-        super().on_run_start(adversary, input, target, model, **kwargs)
+    def on_run_start(
+        self,
+        *,
+        adversary: Optional[torch.nn.Module],
+        input: Optional[Union[torch.Tensor, tuple]],
+        target: Optional[Union[torch.Tensor, Dict[str, Any], tuple]],
+        model: Optional[torch.nn.Module],
```
Comment (dxoigmn) [inline mart/attack/adversary.py:142]: If they're truly optional, then you need to provide a default for keyword arguments. My sense is, thought, you need to remove the `Optional` type.

File: mart/attack/adversary.py [hunk 29]
```diff
@@ -148,8 +159,16 @@ def on_run_start(self, adversary, input, target, model, **kwargs):
...
 
-    def on_run_end(self, adversary, input, target, model, **kwargs):
-        super().on_run_end(adversary, input, target, model, **kwargs)
+    def on_run_end(
+        self,
+        *,
+        adversary: Optional[torch.nn.Module],
+        input: Optional[Union[torch.Tensor, tuple]],
+        target: Optional[Union[torch.Tensor, Dict[str, Any], tuple]],
+        model: Optional[torch.nn.Module],
```
Comment (dxoigmn) [inline mart/attack/adversary.py:168]: same as above

File: mart/attack/adversary.py [hunk 30]
```diff
@@ -162,38 +181,59 @@ def forward(
         self,
         input: Union[torch.Tensor, tuple],
         target: Union[torch.Tensor, Dict[str, Any], tuple],
-        model: torch.nn.Module,
-        **kwargs
+        *,
+        model: Optional[torch.nn.Module],
```
Comment (dxoigmn) [inline mart/attack/adversary.py:185]: Is `model` actually optional? If so, it needs a default. If not, then remove the `Optional` type.

File: mart/attack/adversary.py [hunk 31]
```diff
@@ -162,38 +181,59 @@ def forward(
         self,
         input: Union[torch.Tensor, tuple],
         target: Union[torch.Tensor, Dict[str, Any], tuple],
-        model: torch.nn.Module,
-        **kwargs
+        *,
```
Comment (dxoigmn) [inline mart/attack/adversary.py:184]: Should this be just after `self` so all arguments are keyword arguments?

File: mart/attack/adversary.py [hunk 32]
```diff
@@ -202,10 +242,11 @@ def forward(
...
     def examine(
         self,
-        input: Union[torch.Tensor, tuple],
-        target: Union[torch.Tensor, Dict[str, Any], tuple],
-        model: torch.nn.Module,
-        **kwargs
+        *,
+        input: Optional[Union[torch.Tensor, tuple]],
+        target: Optional[Union[torch.Tensor, Dict[str, Any], tuple]],
+        model: Optional[torch.nn.Module],
```
Comment (dxoigmn) [inline mart/attack/adversary.py:248]: same as above

File: mart/attack/adversary.py [hunk 33]
```diff
@@ -240,10 +281,11 @@ def examine(
...
     def advance(
         self,
-        input: Union[torch.Tensor, tuple],
-        target: Union[torch.Tensor, Dict[str, Any], tuple],
-        model: torch.nn.Module,
-        **kwargs
+        *,
+        input: Optional[Union[torch.Tensor, tuple]],
+        target: Optional[Union[torch.Tensor, Dict[str, Any], tuple]],
+        model: Optional[torch.nn.Module],
```
Comment (dxoigmn) [inline mart/attack/adversary.py:287]: same as above

File: mart/attack/callbacks/progress_bar.py [hunk 34]
```diff
@@ -15,10 +15,11 @@
 class ProgressBar(Callback):
     """Display progress bar of attack iterations with the gain value."""
 
-    def on_run_start(self, adversary, input, target, model, **kwargs):
+    def on_run_start(self, *, adversary, **kwargs):
         self.pbar = tqdm.tqdm(total=adversary.max_iters, leave=False, desc="Attack", unit="iter")
 
-    def on_examine_end(self, adversary, input, target, model, **kwargs):
+    def on_examine_end(self, *, adversary, **kwargs):
+        input = kwargs["input"]
```
Comment (dxoigmn) [inline mart/attack/callbacks/progress_bar.py:22]: Just change this to:
```python
def on_examine_end(self, *, input, adversary, **kwargs):
```
You require input so just make it an parameter of the function!

**Review by dxoigmn — CHANGES_REQUESTED** [review]
More type changes. We really need a mypy test to actually exercise these types.

File: mart/attack/adversary.py (hunk 28 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Removed.

File: mart/attack/adversary.py (hunk 29 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Removed.

File: mart/attack/adversary.py (hunk 30 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Removed.

File: mart/attack/adversary.py (hunk 31 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Somehow `input` and `target` are made non-keyword arguments in `Adversary`. But it should be fine to make them all required keyword arguments in`IterativeGenerator` because we explicitly make that call in `Adversary` with all keyword arguments.

File: mart/attack/adversary.py (hunk 33 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Removed Optional.

File: mart/attack/callbacks/progress_bar.py (hunk 34 above)
Comment (mzweilin) [reply mart/attack/callbacks/progress_bar.py]: Fixed.

File: mart/attack/adversary.py (hunk 31 above)
Comment (dxoigmn) [reply mart/attack/adversary.py]: Is it worth changing `Adversary` then?

Comment (mzweilin) [reply mart/attack/adversary.py:185]: I wouldn't change the specification of `Adversary` in this PR, because it's not backward compatible and it would require changing configurations of Adversary in sequence accordingly.

**Review by dxoigmn — APPROVED** [review]
LGTM!
